File: packages/queue-manager/src/strategies/fairness/rate-limiter.lua
-- Atomic token bucket rate limiter (per shop).
--
-- KEYS[1] = bucketKey
--
-- ARGV[1] = nowMs
-- ARGV[2] = costToConsume
-- ARGV[3] = maxTokens
-- ARGV[4] = refillPerSecond
-- ARGV[5] = ttlMs (optional)
--
-- Returns array:
-- [1] allowed (0/1)
-- [2] delayMs
-- [3] tokensRemaining (after possible consumption)
-- [4] tokensNow (before consumption, after refill)

local bucketKey = KEYS[1]

local nowMs = tonumber(ARGV[1])
local cost = tonumber(ARGV[2])
local maxTokens = tonumber(ARGV[3])
local refillPerSecond = tonumber(ARGV[4])
local ttlMs = tonumber(ARGV[5])

if not nowMs then
  return {0, 60000, 0, 0}
end

if not cost or cost < 0 then
  cost = 0
end

if not maxTokens or maxTokens <= 0 then
  maxTokens = cost
end

if not refillPerSecond or refillPerSecond < 0 then
  refillPerSecond = 0
end

local lastMs = tonumber(redis.call('HGET', bucketKey, 'ts'))
local tokens = tonumber(redis.call('HGET', bucketKey, 'tokens'))

if not lastMs then
  lastMs = nowMs
end

if not tokens then
  tokens = maxTokens
end

-- Refill based on elapsed time
local deltaMs = nowMs - lastMs
if deltaMs < 0 then
  deltaMs = 0
end

local refill = (deltaMs / 1000.0) * refillPerSecond
if refill > 0 then
  tokens = tokens + refill
end

if tokens > maxTokens then
  tokens = maxTokens
end

local tokensNow = tokens
local allowed = 0
local delayMs = 0

if tokens >= cost then
  allowed = 1
  tokens = tokens - cost
else
  allowed = 0
  if refillPerSecond <= 0 then
    delayMs = 60000
  else
    local deficit = cost - tokens
    delayMs = math.ceil((deficit / refillPerSecond) * 1000)
  end
end

redis.call('HSET', bucketKey, 'ts', nowMs, 'tokens', tokens)

if ttlMs and ttlMs > 0 then
  redis.call('PEXPIRE', bucketKey, ttlMs)
end

return {allowed, delayMs, tokens, tokensNow}

File: packages/queue-manager/src/strategies/fairness/rate-limiter.ts
import type { Redis } from 'ioredis';
import { readFileSync } from 'node:fs';
import { metrics } from '@opentelemetry/api';

// F4.4.4: Rate limiter metrics
const meter = metrics.getMeter('neanelu-shopify.rate-limiter');
const rateLimitAllowedTotal = meter.createCounter('ratelimit_allowed_total', {
  description: 'Total requests allowed by rate limiter',
});
const rateLimitDeniedTotal = meter.createCounter('ratelimit_denied_total', {
  description: 'Total requests denied by rate limiter',
});
const rateLimitDelayHistogram = meter.createHistogram('ratelimit_delay_seconds', {
  description: 'Delay imposed by rate limiter in seconds',
  unit: 's',
});

export type RateLimiterParams = Readonly<{
  bucketKey: string;
  nowMs?: number;
  costToConsume: number;
  maxTokens: number;
  refillPerSecond: number;
  /** Expire idle buckets (ms). Default: 10 minutes. */
  ttlMs?: number;
}>;

export type RateLimiterResult = Readonly<{
  allowed: boolean;
  delayMs: number;
  tokensRemaining: number;
  tokensNow: number;
}>;

const DEFAULT_TTL_MS = 10 * 60 * 1000;

const scriptUrl = new URL('./rate-limiter.lua', import.meta.url);
const scriptSource = readFileSync(scriptUrl, 'utf8');

const shaByRedis = new WeakMap<Redis, string>();

async function evalRateLimiter(
  redis: Redis,
  args: {
    bucketKey: string;
    nowMs: number;
    costToConsume: number;
    maxTokens: number;
    refillPerSecond: number;
    ttlMs: number;
  }
): Promise<[number, number, number, number]> {
  const existingSha = shaByRedis.get(redis);

  const keysCount = 1;
  const argv = [
    String(args.nowMs),
    String(args.costToConsume),
    String(args.maxTokens),
    String(args.refillPerSecond),
    String(args.ttlMs),
  ];

  if (existingSha) {
    try {
      const raw = await redis.evalsha(existingSha, keysCount, args.bucketKey, ...argv);
      if (!Array.isArray(raw) || raw.length < 4) throw new Error('rate_limiter_invalid_lua_result');
      return [Number(raw[0]), Number(raw[1]), Number(raw[2]), Number(raw[3])];
    } catch (err) {
      // NOSCRIPT => fall through to load.
      const message = err instanceof Error ? err.message : '';
      if (!message.includes('NOSCRIPT')) throw err;
    }
  }

  const loaded = await redis.script('LOAD', scriptSource);
  if (typeof loaded !== 'string') throw new Error('rate_limiter_invalid_script_sha');
  const sha = loaded;
  shaByRedis.set(redis, sha);

  const raw = await redis.evalsha(sha, keysCount, args.bucketKey, ...argv);
  if (!Array.isArray(raw) || raw.length < 4) throw new Error('rate_limiter_invalid_lua_result');
  return [Number(raw[0]), Number(raw[1]), Number(raw[2]), Number(raw[3])];
}

export async function checkAndConsumeCost(
  redis: Redis,
  params: RateLimiterParams
): Promise<RateLimiterResult> {
  const nowMs = params.nowMs ?? Date.now();
  const ttlMs = params.ttlMs ?? DEFAULT_TTL_MS;

  const bucketKey = params.bucketKey.trim();
  if (!bucketKey) throw new Error('rate_limiter_bucket_key_empty');

  const costToConsume = Math.max(0, Math.floor(params.costToConsume));
  const maxTokens = Math.max(0, Math.floor(params.maxTokens));
  const refillPerSecond = Number(params.refillPerSecond);

  const [allowed, delayMs, tokensRemaining, tokensNow] = await evalRateLimiter(redis, {
    bucketKey,
    nowMs,
    costToConsume,
    maxTokens,
    refillPerSecond,
    ttlMs,
  });

  const isAllowed = allowed === 1;
  const finalDelayMs = Math.max(0, Number(delayMs) || 0);

  // F4.4.4: Emit metrics
  const labels = { bucket_key: bucketKey };
  if (isAllowed) {
    rateLimitAllowedTotal.add(1, labels);
  } else {
    rateLimitDeniedTotal.add(1, labels);
  }
  if (finalDelayMs > 0) {
    rateLimitDelayHistogram.record(finalDelayMs / 1000, labels);
  }

  return {
    allowed: isAllowed,
    delayMs: finalDelayMs,
    tokensRemaining: Math.max(0, Number(tokensRemaining) || 0),
    tokensNow: Math.max(0, Number(tokensNow) || 0),
  };
}
